### PR TITLE
Submodules; and initial protoc-based test

### PR DESCRIPTION
**Pull request by chrisdunelm**
Mostly project configuration, plus a very small amount of generator code. Submodules added to common-protos and protobuf. A single xunit test to check the ability to run protoc.

The next step is to configure CI and ensure this test passes.

### CODE REVIEW

**Comment by chrisdunelm** [comment]
travis has passed on my fork, I'll enable it correctly on this repo shortly...
https://travis-ci.org/chrisdunelm/gapic-generator-csharp

**Comment by chrisdunelm** [comment]
Travis should now be enabled properly, but probably needs to wait for another commit/PR before it'll activate. I'll check on the next PR.

File: Google.Api.Generator/ProtoUtils/ProtoExtensions.cs [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+using System.Linq;
+using System.Reflection;
+
+namespace Google.Api.Generator.ProtoUtils
+{
+    static class ProtoExtensions
+    {
+        public static string CSharpNamespace(this FileDescriptor desc)
+        {
+            // There is no way of reading the C# namespace without using reflection.
```
Comment (jskeet) [inline Google.Api.Generator/ProtoUtils/ProtoExtensions.cs]: Ouch, did I know about this? Maybe I did. Ick - sorry about that.

File: Google.Api.Generator/Google.Api.Generator.csproj [hunk 2]
```diff
@@ -1,8 +1,13 @@
...
+﻿<Project Sdk="Microsoft.NET.Sdk">
 
   <PropertyGroup>
     <OutputType>Exe</OutputType>
     <TargetFramework>netcoreapp2.1</TargetFramework>
+    <LangVersion>latest</LangVersion>
   </PropertyGroup>
 
+  <ItemGroup>
+    <ProjectReference Include="..\protobuf\csharp\src\Google.Protobuf\Google.Protobuf.csproj" />
```
[the remark this answers is not in the document]
Comment (jskeet) [reply Google.Api.Generator/Google.Api.Generator.csproj]: Ah, looks like it was September 22nd, and it didn't make the 3.6 branch.

File: Google.Api.Generator/Generator.cs [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+using Google.Api.Generator.ProtoUtils;
+using Google.Protobuf;
+using Google.Protobuf.Reflection;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Api.Generator
+{
+    static class Generator
```
[the remark this answers is not in the document]
Comment (chrisdunelm) [reply Google.Api.Generator/Generator.cs]: I noticed this, and expected the compiler to give an error, but it didn't so I left it ;) Will fix.

File: Google.Api.Generator/Google.Api.Generator.csproj (hunk 2 above)
[the remark this answers is not in the document]
Comment (chrisdunelm) [reply Google.Api.Generator/Google.Api.Generator.csproj]: ack

File: Google.Api.Generator/Generator.cs (hunk 3 above)
[the remark this answers is not in the document]
Comment (chrisdunelm) [reply Google.Api.Generator/Generator.cs]: Both done.

File: Google.Api.Generator/ProtoUtils/ProtoCatalog.cs [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+using Google.Protobuf.Reflection;
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Api.Generator.ProtoUtils
+{
+    /// <summary>
+    /// A catalog of all loaded proto messages.
+    /// </summary>
+    class ProtoCatalog
```
[the remark this answers is not in the document]
Comment (chrisdunelm) [reply Google.Api.Generator/ProtoUtils/ProtoCatalog.cs]: Done.

File: Google.Api.Generator/ProtoUtils/ProtoExtensions.cs (hunk 1 above)
Comment (chrisdunelm) [reply Google.Api.Generator/ProtoUtils/ProtoExtensions.cs]: Yes, we discussed this a few months ago. This is fine for the time being - it's nicely contained, this is the only place this is required.
If/when protobuf is changed I'll update this.

**Comment by chrisdunelm** [comment]
Travis is failing because it says it cannot find the 2.1.500 sdk, which is odd as when I ran travis on my fork it worked and the test passed.
If it's OK with you, I'd like to ignore this for this PR, and I'll work out what's going wrong in a (near-)future PR.

**Comment by chrisdunelm** [comment]
Looks like this may be a travis issue: https://travis-ci.community/t/error-today-when-running-net-core-builds/1148/4
So I won't spend any more time today looking into it.

**Pull request merged**